WIP Add Cluster Wide SELinux Capability

swagiaal opened the pull request:
- Add flag to api-server to enable capability
- Ensure pod spec does not contain SELinux context if capability is
  not on
- Add flag to kubelet to enable capability
- Ensure node has SELinux enabled if capability is on.
- Ensure that pod spec does not contain SELinux context if capability
  is not on.

Commit:
WIP Add Cluster Wide SELinux Capability

- Add flag to api-server to enable capability
- Ensure pod spec does not contain SELinux context if capability is
  not on
- Add flag to kubelet to enable capability
- Ensure node has SELinux enabled if capability is on.
- Ensure that pod spec does not contain SELinux context if capability
  is not on.

pmorie commented on cmd/kubelet/app/server.go:
```diff
@@ -69,6 +69,7 @@ const defaultRootDir = "/var/lib/kubelet"
 type KubeletServer struct {
 	Address                        net.IP
 	AllowPrivileged                bool
+	enableSELinuxIntegration       bool
```
Should be an exported field; `EnableSELinuxIntegration`

dchen1107 replied:
+1

swagiaal replied:
addressed

pmorie commented on cmd/kubelet/app/server.go:
```diff
@@ -779,6 +782,7 @@ func makePodSourceConfig(kc *KubeletConfig) *config.PodConfig {
 type KubeletConfig struct {
 	Address                        net.IP
 	AllowPrivileged                bool
+	enableSELinuxIntegration       bool
```
Should also be exported.

swagiaal replied:
addressed

pmorie commented:
@pweil- @dchen1107 PTAL

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -1921,6 +1921,10 @@ func ValidateSecurityContext(sc *api.SecurityContext) errs.ValidationErrorList {
 	}
+
+	if sc.SELinuxOptions != nil && !capabilities.Get().EnableSELinuxIntegration {
+		allErrs = append(allErrs, errs.NewFieldForbidden("selinuxOptions", "Cannot specify selinuxOptions if SELinux integration is not enabled"))
```
Typo, the field is named `seLinuxOptions` in json/yaml

pweil- replied:
Why is this an error condition in both validation and the kubelet.  If I submit a pod with seLinuxOptions and the system isn't enforcing does it matter?  Shouldn't my pod run anyway since the system is less restrictive?

pmorie replied:
@pweil- We need a validation in the Kubelet to catch existing pods, afaik.  I'd love to be wrong.  Thoughts?

pweil- replied:
I'm not questioning if it belongs in both places.  If we want the verification then this is correct.  But I'm wondering if this is an error condition at all.  If I pass a pod with SELinuxOptions and the cluster is not using SELinux why should it fail?  Presumably my pod can still run fine

pmorie replied:
@pweil- Good point, it's supposed to DTRT.  I guess in that case you should get the behavior we have in Kubernetes now, which is to give your emptyDir volumes a usable SELinux context.

pmorie replied:
@smarterclayton @swagiaal Thoughts about that?

swagiaal replied:
@pweil- The presumption here is that if the user specifies an SELinux context then they want the container to run under that context. If SELinux is disabled then the pod will run of course but it will be less restricted than the user desired it to be.

dchen1107 replied:
Kubelet already does validation by using the same validation package.

swagiaal replied:
> Typo, the field is named seLinuxOptions in json/yaml

addressed

> Kubelet already does validation by using the same validation package.

addressed

pmorie commented on pkg/capabilities/capabilities.go:
```diff
@@ -31,6 +31,10 @@ type Capabilities struct {
 	PerConnectionBandwidthLimitBytesPerSec int64
+
+	// EnableSELinuxIntegration controls whether SELinux
+	// integration is expected from Kubernets
```
Typo: Kubernets

swagiaal replied:
addressed

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -2346,10 +2348,13 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 	networkConfigured := kl.doneNetworkConfigure()
 
+	selinuxRequired := capabilities.Get().EnableSELinuxIntegration
+	selinuxConfigured := (selinuxRequired && selinux.SelinuxEnabled()) || !selinuxRequired
```
Unfortunately the libcontainer selinux library doesn't provide a stub for !linux, so you have to abstract this behind a golang build flag.  See `root_context_{linux,unsupported}.go`

pmorie replied:
Btw, I would also call this `selinuxEnabled`.  Configured makes it sound like it has something to do with policy.

pmorie replied:
Actually maybe something like `selinuxOK`

swagiaal replied:
addressed

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -2366,6 +2371,9 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 			messages = append(reasons, "network not configured correctly")
 		}
+		if !selinuxConfigured {
+			messages = append(reasons, "SELinux is required by the cluster but not enabled on this node")
```
I would make this a little more verbose to match the phrasing of the name of the capability:

"SELinux integration is enabled in the cluster but SELinux is not enabled on this node.  See /etc/sysconfig/selinux."

dchen1107 replied:
Sounds like we imply here if API server is enabled with selinux-integration, all nodes must be selinux-integration enabled. Heterogeneous cluster in this case is not allowed. Is that right?

dchen1107 replied:
Otherwise, replace cluster with node in above message?

pmorie replied:
@dchen1107 Heterogenous is not allowed, correct.  I don't know a single person / group that wants it.

dchen1107 replied:
ACK. This means:

* Admins have to decide if a cluster with SELinux or not at the cluster creation time. If they want a different configuration, they have to create a new one.
* If master is configured with SELinux, all Kubelets should configured with SELinux; any Kubelet without such config should be rejected by master by marking as NotReady.
* If master is configured withoutLinux, all Kubelets shouldn't configured with SELinux; any Kubelet with such config should be rejected by master by marking as NotReady. 

In this case, why we need an extra flag in Kubelet here? Why not introducing such to NodeSpec?

pmorie replied:
@dchen1107 

> * Admins have to decide if a cluster with SELinux or not at the cluster creation time. If they want a different configuration, they have to create a new one.

Not quite.  Technically I think it would be possible to turn SELinux integration **on** -- and the system will DTRT so that containers correctly with SELinux
enabled but are not isolated from one another (ie, they have a usable SELinux context for the system policy but no MCS labels).

That said, I think it is not required at all to work out the state transitions.  I don't know of anyone that wants to toggle back and forth between enabled and disabled.
If you run into a policy issue, you can setenforce 0 on the node and SELinux will allow actions that are denied by policy.  Since supporting toggling between SELinux enabled and disabled isn't a priority use-case at all, I'm fine defining the semantics as you suggest.

> * If master is configured with SELinux, all Kubelets should configured with SELinux; any Kubelet without such config should be rejected by master by marking as NotReady.
> * If master is configured withoutLinux, all Kubelets shouldn't configured with SELinux; any Kubelet with such config should be rejected by master by marking as NotReady. 

Yes

> In this case, why we need an extra flag in Kubelet here? Why not introducing such to NodeSpec?

I'm not sure what you mean by an extra flag in the Kubelet.  Elaborate?  And, will NodeSpec cover the nil cloud provider case?

pmorie replied:
Full disclosure, I am still reading up on the mechanics of NodeSpec.

swagiaal replied:
hmm... I will look into NodeSpec

> I would make this a little more verbose to match the phrasing of the name of the capability:
> 
> "SELinux integration is enabled in the cluster but SELinux is not enabled on this node. See /etc/sysconfig/selinux."

Addressed

pmorie commented on pkg/kubelet/util.go:
```diff
@@ -77,6 +77,16 @@ func canRunPod(pod *api.Pod) error {
+	if !capabilities.Get().EnableSELinuxIntegration {
+		for _, container := range pod.Spec.Containers {
+			if securitycontext.HasSELinuxOptions(&container) {
+				return fmt.Errorf("pod with UID %q specifie 'selinuxOptions' but SELinuxIntegration is disabled", pod.UID)
```
Typo: specifies

swagiaal replied:
addressed

pmorie commented on pkg/securitycontext/util.go:
```diff
@@ -47,6 +47,14 @@ func HasCapabilitiesRequest(container *api.Container) bool {
 }
 
+// HasSELinuxOptions returns true if the containers security context specifies SELinux options
+func HasSELinuxOptions(container *api.Container) bool {
```
Remember that there will soon be a pod-level selinux context -- you might want to pass the pod, too.

swagiaal replied:
I removed this function

pweil- commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -207,6 +208,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EtcdPathPrefix, "etcd-prefix", s.EtcdPathPrefix, "The prefix for all resource paths in etcd.")
 	fs.StringSliceVar(&s.CorsAllowedOriginList, "cors-allowed-origins", s.CorsAllowedOriginList, "List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching.  If this list is empty CORS will not be enabled.")
 	fs.BoolVar(&s.AllowPrivileged, "allow-privileged", s.AllowPrivileged, "If true, allow privileged containers.")
+	fs.BoolVar(&s.EnableSELinuxIntegration, "selinux-integration", s.EnableSELinuxIntegration, "If true the api-server will not accept pods which specify SELinux context. [default=false]")
```
Is the wording on this right?  I think you meant "if true will accept..."

pmorie replied:
@pweil- You're right, the wording is incorrect.

dchen1107 replied:
+1

swagiaal replied:
addressed

pmorie commented:
This fixes #14131

pmorie commented:
@swagiaal Are you going to handle EmptyDir in this PR?

dchen1107 commented:
I might misread the pr here. I thought you are going to detect SELinux is enabled or not on node. If it is required for a kubelet from flag --selinux-integration, you are going to mark the node is Not Ready with proper reason. But I didn't see both logics: 1) detects SELinux enabled or not 2) If mismatch, mark the node NotReady in current PR.

Commit:
Address feedback

Commit:
Remove redundant selinuxOptions check from kubelet

Commit:
Abstract selinuxEnabled() behind root_context et al

Commit:
Correct comment for selinux-integration option

swagiaal commented:
> @swagiaal Are you going to handle EmptyDir in this PR?

I will address that in a separate PR

> specified

Corrected

> Kubernetes

Corrected

swagiaal commented:
> I might misread the pr here. I thought you are going to detect SELinux is enabled or not on node. If it is required for a kubelet from flag --selinux-integration, you are going to mark the node is Not Ready with proper reason. But I didn't see both logics: 1) detects SELinux enabled or not 2) If mismatch, mark the node NotReady in current PR.

@dchen1107 that is handled here:
https://github.com/kubernetes/kubernetes/pull/14513/files#diff-bf28da68f62a8df6e99e447c4351122dR2351
and here:

https://github.com/kubernetes/kubernetes/pull/14513/files#diff-bf28da68f62a8df6e99e447c4351122dR2374

dchen1107 commented:
When you said detecting SELinux enabled or not, I thought you are referring to call sestatus or some equivalent commands to make sure SELinux model is enabled and enforced etc. But above two links you list only checked flags, right?

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -2346,10 +2347,13 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 	networkConfigured := kl.doneNetworkConfigure()
 
+	selinuxRequired := capabilities.Get().EnableSELinuxIntegration
+	selinuxOK := (selinuxRequired && kl.selinuxEnabled()) || !selinuxRequired
```
If SELinux integration isn't enabled and SELinux is enabled on the host, I think that should be an error.

swagiaal commented:
> When you said detecting SELinux enabled or not, I thought you are referring to call sestatus or some equivalent commands to make sure SELinux model is enabled and enforced etc. But above two links you list only checked flags, right?

Ah I see what you are saying. 'kl.selinuxEnabled()' uses 'github.com/docker/libcontainer/selinux.SelinuxEnabled()' which in turn checks for the presence of an selinuxfs. The presence of the fs is an indication that SELinux is enabled on the system, but we also need to check that it is "enforcing". I'll add that.

pmorie commented:
@swagiaal We should not have an assertion about enforcing.  Only enabled or not.  If we can't run the Kubelet with SELinux in permissive mode, debugging is going to be a nightmare.

pweil- commented:
> @swagiaal We should not have an assertion about enforcing. Only enabled or not. If we can't run the Kubelet with SELinux in permissive mode, debugging is going to be a nightmare.

Agree.

I am +1 for having a way for masters and nodes to agree that SELinux should or shouldn't be running according to administrator wishes and being able to specify that a node is not ready due to running in an unexpected state. 

I'm still hung up on rejecting a pod that specifies SELinux options when the cluster is running in a non-SELinux enabled configuration.  I don't see a reason it should be an fatal condition, anything that would normally be setting labels would just ignore them based on the cluster settings anyway, I assume. 

Speaking of which, are you expecting drive behavior in things like the security context provider based on this setting?  I don't think I saw that in the pull but you probably want to at least throw some TODO items in where you'll be expecting to drive behavior with this flag.

pmorie commented on pkg/capabilities/capabilities.go:
```diff
@@ -31,6 +31,10 @@ type Capabilities struct {
 	// PerConnectionBandwidthLimitBytesPerSec limits the throughput of each connection (currently only used for proxy, exec, attach)
 	PerConnectionBandwidthLimitBytesPerSec int64
+
+	// EnableSELinuxIntegration controls whether SELinux
```
I think this comment should define the semantics that you can expect from this capability in detail.

thockin replied:
+1.  Also weird linewrap

pmorie commented on pkg/kubelet/root_context_linux.go:
```diff
@@ -33,3 +33,7 @@ func (kl *Kubelet) getRootDirContext() (string, error) {
 	// Get the SELinux context of the rootDir.
 	return selinux.Getfilecon(kl.getRootDir())
 }
+
```
I'm cool with this going into these files for now, but we will probably wind up renaming this file to something like `selinux_linux.go` shortly.

swagiaal replied:
> we will probably wind up renaming this file to something like selinux_linux.go shortly.

+1

> I'm still hung up on rejecting a pod that specifies SELinux options when the cluster is running in a non-SELinux enabled configuration. I don't see a reason it should be an fatal condition, anything that would normally be setting labels would just ignore them based on the cluster settings anyway, I assume.

So lets say I have an untrusted pod. I am worried about running it so I run it under the label
'untrusted_u:untrusted_r:untrusted_t' to which my policy give very limited access. In that case if a the above pod is allowed to run on a system with SELinux not enabled then it goes against what the user wanted

pmorie commented:
> I'm still hung up on rejecting a pod that specifies SELinux options when the cluster is running in a non-SELinux enabled configuration. I don't see a reason it should be an fatal condition, anything that would normally be setting labels would just ignore them based on the cluster settings anyway, I assume.

It's arguable that this falls under DTRT.  The SELinux context is a request.  If the system can't fulfill the request, should it DTRT and run your pods without SELinux confinement, or should it reject the pod?

smarterclayton commented:
I believe very strongly that not specifying an SELinux context means you don't care and accept the defaults of the system.  SELinux is almost always managed by the system, is not an optional behavior, and cannot be changed by end users.  If you *must* or *want* to run in a specific SELinux context, that's what it means when you specify the value.  When you do that, your admin must decide whether to allow that (because of a policy).  Running unconfined is still an affirmative label (you say you want to run as unconfined / system) by specifying a value for label.

pmorie commented:
@smarterclayton Sounds like the matrix looks like

| Pod requires SELinux? | SELinux capability enabled? | Behavior                                                                                                                                  |
|-----------------------|-----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|
| yes                   | yes                         | allow admin to reject w/ admission controller                                                                                             |
| yes                   | no                          | reject                                                                                                                                    |
| no                    | yes                         | correctly leverage SELinux but don't isolate containers from one another using SELinux (similar to how SELinux works today in Kubernetes) |
| no                    | no                          | accept                                                                                                                                    |

@swagiaal @pweil- @dchen1107

smarterclayton commented:
I don't like the second row.  For one reason, it means pod definitions that
conditionally set a SELinux label fail on certain systems.  How does that
benefit the user?  If you ask for a label, and the system doesn't support
labels, the system also isn't enforcing labels, so your label ask has no
meaning.  Failing seems harsh.

On Fri, Sep 25, 2015 at 3:58 PM, Paul Morie <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> Sounds like the
> matrix looks like
> Pod requires SELinux? SELinux capability enabled? Behavior yes yes allow
> admin to reject w/ admission controller yes no reject no yes correctly
> leverage SELinux but don't isolate containers from one another using
> SELinux (similar to how SELinux works today in Kubernetes) no no accept
>
> @swagiaal <https://github.com/swagiaal> @pweil-
> <https://github.com/pweil-> @dchen1107 <https://github.com/dchen1107>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143337611>
> .
>

pmorie commented:
@smarterclayton so, under DTRT, just accept the pods and run them as you would normally with SELinux not enabled?

smarterclayton commented:
I don't think any pod *requires* SELinux.  I think some pods want to run
with specific labels.  If they depend on those labels or selinux enabled in
code they're going to fail anyway (in some cases).  I think that could be
optional admission policy, but doesn't seem like a standard default.

On Fri, Sep 25, 2015 at 4:59 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I don't like the second row.  For one reason, it means pod definitions
> that conditionally set a SELinux label fail on certain systems.  How does
> that benefit the user?  If you ask for a label, and the system doesn't
> support labels, the system also isn't enforcing labels, so your label ask
> has no meaning.  Failing seems harsh.
>
> On Fri, Sep 25, 2015 at 3:58 PM, Paul Morie <notifications@github.com>
> wrote:
>
>> @smarterclayton <https://github.com/smarterclayton> Sounds like the
>> matrix looks like
>> Pod requires SELinux? SELinux capability enabled? Behavior yes yes allow
>> admin to reject w/ admission controller yes no reject no yes correctly
>> leverage SELinux but don't isolate containers from one another using
>> SELinux (similar to how SELinux works today in Kubernetes) no no accept
>>
>> @swagiaal <https://github.com/swagiaal> @pweil-
>> <https://github.com/pweil-> @dchen1107 <https://github.com/dchen1107>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143337611>
>> .
>>
>
>

pmorie commented:
@smarterclayton Optional admission policy to reject pods that request SELinux options?  That seems reasonable to me.

smarterclayton commented:
Although I don't even think it's relevant enough to implement.  I can't
imagine the value to an operator or developer

On Sep 25, 2015, at 5:02 PM, Paul Morie <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> Optional admission
policy to reject pods that request SELinux options? That seems reasonable
to me.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143354076>
.

smarterclayton commented:
If the admin does not have selinux, or has intentionally disabled selinux,
I don't think they care that a pod specified it expected to be run under a
label.  The fact that the value is ignored is an intentional choice from
the admin (no selinux).  I think is roughly equivalent to a tar extract
operation where selinux is disabled.  Should tar fail and tell you that
your labels won't have any effect?  A super conscious admin might.
Everyone else will just hate tar a little bit more.

On Sep 25, 2015, at 5:01 PM, Paul Morie <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> so, under DTRT, just
accept the pods and run them as you would normally with SELinux not enabled?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143353847>
.

thockin commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -207,6 +208,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EtcdPathPrefix, "etcd-prefix", s.EtcdPathPrefix, "The prefix for all resource paths in etcd.")
 	fs.StringSliceVar(&s.CorsAllowedOriginList, "cors-allowed-origins", s.CorsAllowedOriginList, "List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching.  If this list is empty CORS will not be enabled.")
 	fs.BoolVar(&s.AllowPrivileged, "allow-privileged", s.AllowPrivileged, "If true, allow privileged containers.")
+	fs.BoolVar(&s.EnableSELinuxIntegration, "selinux-integration", s.EnableSELinuxIntegration, "If true the api-server will accept pods which specify SELinux context. [default=false]")
```
The word "integration" adds nothing to this flag or the variables.  Shorter is better.  It is "Go" after all and not "GooglesNewProgrammingLanguage"

Commit:
Add a flag to NodeSpec to indicate SELinux status

swagiaal commented:
@dchen1107 Re Using NodeSpec did you mean something like this:
https://github.com/kubernetes/kubernetes/commit/28b108a994d909eaf3aa1831541172428ee4d04a

swagiaal commented:
> If the admin does not have selinux, or has intentionally disabled selinux, I don't think they care that a pod specified it expected to be run under a label. The fact that the value is ignored is an intentional choice from the admin (no selinux).

Hmm.. Is there value in having this flag then ? We can achieve the above by making all SELinux operations conditional on whether SELinux is enabled on the node or not.

thockin commented:
Do we need to step back from the code and write a proposal doc first?

pmorie commented:
I think we can just omit the kubelet flag for now.  The cluster capability is sufficient to drive the kubelet behaviors.

swagiaal commented:
> Do we need to step back from the code and write a proposal doc first?

Hmm... I can do that. I thought this was simple enough that the issue was enough, but it is proving otherwise.

> I think we can just omit the kubelet flag for now. The cluster capability is sufficient to drive the kubelet behaviors.

https://github.com/kubernetes/kubernetes/commit/28b108a994d909eaf3aa1831541172428ee4d04a eliminates the kubelet flag already. My questions is regarding the api-server flag and the capability. I think we can achieve what @smarterclayton is describing by just gating SELinux operations on whether SELinux is enabled on the node.

pmorie commented:
I think we can pull the API server flag too.

pmorie commented:
Actually, I retract the last comment.

pmorie commented:
@smarterclayton So do you think we even need this capability PR at all?

pmorie commented:
After thinking about it a little more today, perhaps this is what we really want:

1.  A setting (not sure if this is really a capability or what it's a setting _of_) that allows operators to express that all nodes should have SELinux **enabled** (enforcing or permissive) on a node.  If this setting is enabled, and a SELinux is not enabled on a node, that node should report `NotReady`.
2.  If SELinux is enabled on a node (regardless of the above setting), the system will treat SELinux as follows:
    1.  If a container does not specify an SELinux context, the system will attempt to run your containers with a usable SELinux context but won't attempt to isolate pods from one another with SELinux.  A sensible behavior for the docker runtime is to delegate to docker -- the daemon will give you a random MCS label per container.  This is basically the level of SELinux support you get today if you don't specify an SELinux context for a container.
    2.  If a container does specify an SELinux context, the container is run with that SELinux context

@smarterclayton @swagiaal @dchen1107 @thockin

dchen1107 commented:
I like your new proposal at https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143946903. It is way cleaner than what we are having here by introducing flag to both master and nodes. But I have following question:

* How to deal with node auto register? The newly joining node will get this capacity or whatever from somewhere, such as API server or config store (?), then check if SELinux enabled or not, then claim its Readiness?

smarterclayton commented:
I'm concerned that we would apply node policy in two places.  Today, we
clearly state that nodes control their policy.  An admin that wants to
guarantee selinux is on for all their nodes has plenty of tools outside the
platform to ensure that.  I don't necessarily think that we need a tool in
the master to drive that.

I would suggest simply dropping 1. in favor of 2.  1. is a deployment
problem, and when we have centralized node config via the other "store
config in the apiserver" proposals, the problem goes away.


On Sep 29, 2015, at 9:14 AM, Dawn Chen <notifications@github.com> wrote:

I like your new proposal at #14513 (comment)
<https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143946903>.
It is way cleaner than what we are having here by introducing flag to both
master and nodes. But I have following question:

   - How to deal with node auto register? The newly joining node will get
   this capacity or whatever from somewhere, such as API server or config
   store (?), then check if SELinux enabled or not, then claim its Readiness?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14513#issuecomment-143969657>
.

swagiaal commented:
Alright, unless there are any objections, I am going to close this PR in favor of just using node SELinux settings to enable/disable SELinux integration.
@smarterclayton @pmorie

> How to deal with node auto register? The newly joining node will get this capacity or whatever from somewhere, such as API server or config store (?), then check if SELinux enabled or not, then claim its Readiness?

@dchen1107 https://github.com/kubernetes/kubernetes/commit/28b108a994d909eaf3aa1831541172428ee4d04a removes the flag from the kubelet. The node reports the SELinux status to the master and the master can reject the node if it does not match the desired setting. That is just for clarification though since I am closing this PR :)

pmorie commented:
No objection.
